xdsl/dialects/builtin.py

```diff
@@ -1005,13 +1005,45 @@ def unpack(
 
 
 @irdl_attr_definition
-class ComplexType(ParametrizedAttribute, BuiltinAttribute, TypeAttribute):
+class ComplexType(StructPackableType[complex], ParametrizedAttribute, BuiltinAttribute, TypeAttribute):
```

Comment:
To me, it feels like the "correct" way to do this is to make this a `StructPackableType[int | float]` and just pack alternating real/imaginary parts and then override each of the required functions.